### ISSUE TITLE
plugin/prometheus No metrics returned

### ISSUE THREAD

**Issue by Nemric** [coredns/coredns#3688]
What happened :
Running Kubernetes 1.17.3 initialized with Kubeadm, in testing environment, 1 master and 1 worker node.
I don't have metrics for one of my coredns pod, the one on master node, and it appears to be "down" in prometheus

What you expected to happen:
Having metrics for this pod like the one running on worker node

How to reproduce it (as minimally and precisely as possible):

Command from a container in the cluster : 

On Master node :

```
curl http://192.168.132.27:9153/metrics -v
*   Trying 192.168.132.27:9153...
* TCP_NODELAY set
* Connected to 192.168.132.27 (192.168.132.27) port 9153 (#0)
> GET /metrics HTTP/1.1
> Host: 192.168.132.27:9153
> User-Agent: curl/7.66.0
> Accept: */*
> 
```
and after minutes 

```
* Recv failure: Connection reset by peer
* Closing connection 0
curl: (56) Recv failure: Connection reset by peer
```

On my Worker node : 

```
curl http://192.168.203.57:9153/metrics -v
*   Trying 192.168.203.57:9153...
* TCP_NODELAY set
* Connected to 192.168.203.57 (192.168.203.57) port 9153 (#0)
> GET /metrics HTTP/1.1
> Host: 192.168.203.57:9153
> User-Agent: curl/7.66.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Content-Type: text/plain; version=0.0.4; charset=utf-8
< Date: Fri, 21 Feb 2020 10:04:23 GMT
< Transfer-Encoding: chunked
< 
# HELP coredns_build_info A metric with a constant '1' value labeled by version, revision, and goversion from which CoreDNS was built.
# TYPE coredns_build_info gauge
coredns_build_info{goversion="go1.13.4",revision="c2fd1b2",version="1.6.5"} 1
# HELP coredns_cache_hits_total The count of cache hits.
# TYPE coredns_cache_hits_total counter
...
# HELP process_virtual_memory_max_bytes Maximum amount of virtual memory available in bytes.
# TYPE process_virtual_memory_max_bytes gauge
process_virtual_memory_max_bytes -1
* Connection #0 to host 192.168.203.57 left intact
```

Environment:

- the version of CoreDNS: 1.6.5
- Corefile: 
`.:53 {
    errors
    health {
       lameduck 5s
    }
    ready
    kubernetes cluster.local in-addr.arpa ip6.arpa {
       pods insecure
       fallthrough in-addr.arpa ip6.arpa
       ttl 30
    }
    prometheus :9153
    forward . /etc/resolv.conf
    cache 30
    loop
    reload
    loadbalance
}`

- OS (e.g: `cat /etc/os-release`): 
`PRETTY_NAME="Debian GNU/Linux 10 (buster)"`

**Comment by chrisohaver** [comment]
> curl: (56) Recv failure: Connection reset by peer

In this case, the "peer" could be a firewall.  Is there a firewall running on the master node?

**Comment by Nemric** [comment]
There's no firewall, as you can see, curl http://x.x.x.x:9153/metrics is successfully connected to coredns but don't get any metrics...
I think that http request is handled but prometheus plugin don't send metrics ...

I don't really know what is the "peer" you mentionned, perhaps you mean that I called "a container in the cluster" which is in this case a fedora .... I tried to with any container that accept "kubectl exec -ti xxx /bin/bash" (from memory ^^, note that Coredns won't accept this command) and have curl or wget for testing connectivity I prefer curl for that ...

**Comment by chrisohaver** [comment]
"peer" from the curl error message "Connection reset by _peer_".

Anything in the CoreDNS logs (of the affected Pod)?

**Comment by Nemric** [comment]
No relevant logs ...
```
Loading....:53
[INFO] plugin/reload: Running configuration MD5 = 4e235fcc3696966e76816bcd9034ebc7
CoreDNS-1.6.5
linux/amd64, go1.13.4, c2fd1b2
```

Http Connection is reseted after minutes of "no responses" ... nothing relevant here in my opinion ...

**Comment by chrisohaver** [comment]
Can you reach http services on other Pods running on the master node?

**Comment by chrisohaver** [comment]
A couple more things to try:

1. Can you curl the master node coredns's health/readiness ports (8080 and 8081) from a container in the in the cluster.
2. Can you curl the prometheus port from the master node itself (not from a container).

**Comment by Nemric** [comment]
Sorry forgotten to mention that in my question, I did it, so here are today's results : 

```
# curl http://192.168.203.57:8080/health -v
*   Trying 192.168.203.57:8080...
* TCP_NODELAY set
* Connected to 192.168.203.57 (192.168.203.57) port 8080 (#0)
> GET /health HTTP/1.1
> Host: 192.168.203.57:8080
> User-Agent: curl/7.66.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Date: Fri, 21 Feb 2020 15:27:52 GMT
< Content-Length: 2
< Content-Type: text/plain; charset=utf-8
< 
* Connection #0 to host 192.168.203.57 left intact
OK
```

```
# curl http://192.168.203.57:8181/ready -v
*   Trying 192.168.203.57:8181...
* TCP_NODELAY set
* Connected to 192.168.203.57 (192.168.203.57) port 8181 (#0)
> GET /ready HTTP/1.1
> Host: 192.168.203.57:8181
> User-Agent: curl/7.66.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Date: Fri, 21 Feb 2020 15:28:57 GMT
< Content-Length: 2
< Content-Type: text/plain; charset=utf-8
< 
* Connection #0 to host 192.168.203.57 left intact
OK
```

**Comment by Nemric** [comment]
Sorry, tested the wrong pod ^^
but results are the same on "buggy" pod :
```
# curl http://192.168.132.27:8080/health
OK
```
```
# curl http://192.168.132.27:8181/ready
OK
```

**Comment by chrisohaver** [comment]
Confounding.  Grasping at straws here, try replacing `error` with `debug` in your Corefile.  This prompts _some_ plugins to log more verbosely (but unfortunately I don't think `prometheus` is one of them).  It also disables auto crash recovery, which could be obfuscating some kind of unexpected error.

**Comment by Nemric** [comment]
I add `debug` before `errors` in my core file so it looks like that now :

```
.:53 {
    debug 
    errors
    health {
       lameduck 5s
    }
    ready
    kubernetes cluster.local in-addr.arpa ip6.arpa {
       pods insecure
       fallthrough in-addr.arpa ip6.arpa
       ttl 30
    }
    prometheus :9153
    forward . /etc/resolv.conf
    cache 30
    loop
    reload
    loadbalance
}
```
I did run  this command `kubectl rollout restart deployment coredns -n kube-system` to reload conf...

But logs are still not really verbose ...

`Loading....:53
[INFO] plugin/reload: Running configuration MD5 = 9bdd6d5b1e52f2e8db00ef5eab058723
CoreDNS-1.6.5
linux/amd64, go1.13.4, c2fd1b2`

Before you ask me, I ran`nslookup google.com 192.168.132.27` and this pod did answer my dns request well...

**Comment by chrisohaver** [comment]
> I add `debug` before `errors` in my core file so it looks like that now :

It's crucial to _replace_ errors with debug (IOW, remove error and add debug).  Using them both at the same time nullifies the disabling of crash recovery (including errors re-enables it). 

> But logs are still not really verbose ...

That's expected. Most plugins, including prometheus do not produce verbose logs.  However, debug also disables auto crash recovery which if disabled would more clearly tell us whether or not CoreDNS is crashing and auto recovering when it accepts a Prometheus connection.  Disabling auto crash recovery is in fact the primary function of the debug plugin, verbose logs is secondary. https://github.com/coredns/coredns/blob/master/plugin/debug/README.md

**Comment by Nemric** [comment]
So I did what you said when you said it : 
> It's crucial to replace errors with debug

and today morning ... no relevant log
```
 Loading....:53
[INFO] plugin/reload: Running configuration MD5 = 766c0fd28e5bec30977ef4b780048f3c
CoreDNS-1.6.5
linux/amd64, go1.13.4, c2fd1b2
```

**Comment by miekg** [comment]
If you can replicate this without kubernetes, I consider it a bug.

**Comment by chrisohaver** [comment]
My inclination is that this is somehow environmental.  It works fine on the worker node, but not the master.  CoreDNS binaries are identical, using same configuration, receiving same request, such that they will be processing the request and responding in the same way.  It suggests that the response is somehow getting lost in transit on the master.

Sniffing traffic on the master node's CoreDNS Pod should verify if this is the case or not.  https://medium.com/@xxradar/how-to-tcpdump-effectively-in-kubernetes-part-1-a1546b683d2f

**Comment by Nemric** [comment]
Sorry, I won't be able to take more time for this "bug" (or not) as it could be environment problem and I've done my best ...
It would have been easier to go in "terminal" in coredns container to make some internal checks before make any tcpdump ....
I'll perhaps give it a new try later but some other dev wait for me ...
You can close bug for the moment (don't know how to do it ^^)

**Comment by chrisohaver** [comment]
OK thanks, please reopen if you have more time to come back to it.

**Comment by QuinnBast** [comment]
I am also experiencing this...

I cannot hit the `/metrics` endpoint if I use the coredns service in k8s. However, if I manually port-forward the coredns pods port 9153 and `curl` from my local machine I can get a response. But calling this endpoint between pods ends up hanging forever and not resolving.

Edit: I found my issue. I had a `NetworkPolicy` that was getting automatically applied from RKE2 which blocked traffic to the metrics port (9153). I had to modify the network policy that ships with RKE2 out of the box to allow the 9153 port to talk to coredns
